### PR TITLE
Add sdpa and fa2 the Wav2vec2 family. 

### PR DESCRIPTION
**Pull request by kamilakesbi**
Co-authored-by: @kamilakesbi <kamil@huggingface.co>
Co-authored-by: @jp1924 <jp42maru@gmail.com>

# What does this PR do?

This PR aims at solving issue #30073 by adding [SPDA](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) and [Flash Attention 2](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) to the  Wav2Vec2 modelling code.

@jp1924  has already done most of the necessary changes [here](https://github.com/huggingface/transformers/issues/30073#issuecomment-2042025148). Based on his code, I added SDPA and made sure it passed `make fixup`  and updated the documentation.

Next steps: 
- Add Flash Attention 2. @jp1924 feel free to participate! 
- Add fast and slow tests as suggested in #30073

## Who can review?

cc @sanchit-gandhi

### CODE REVIEW

**Comment by jp1924** [comment]
@kamilakesbi 
I'd like to contribute too but I can't commit because I don't have permission, do you know what I can do?

**Comment by jp1924** [comment]
Sure! I'm new to co-authored-by and didn't know what that meant, so thanks for clearing that up!

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (sanchit-gandhi) [inline tests/models/wav2vec2/test_modeling_wav2vec2.py:2005]: Currently, this test looks more like a fast test in terms of what it checks (that the logits from fa2 match those from eager). I would move it into the fast tests and use the dummy models as done here: https://github.com/huggingface/transformers/blob/9b5a6450d481b0f02834684ffd8b3ba4cbbd6fe0/tests/models/whisper/test_modeling_whisper.py#L891 

Note that if we load two full-size models and do two forward passes, the slow test becomes un-necessarily heavy. Instead, we can use the transcription from the "eager" model as the expected transcription (or ground truth), and then simply check that the transcription from the "fa2" model matches what we expect.

More specifically, I would copy the existing test from: https://github.com/huggingface/transformers/blob/7afade208695f1a25dcc9656d553d51ace67264e/tests/models/wav2vec2/test_modeling_wav2vec2.py#L1481

But update it to use `attn_implementation="flash_attention_2"`, confirming that the transcription result is the same as the current `EXPECTED_TRANSCRIPTIONS`

=> this is good enough to confirm the model works with FA2 with inference

Comment (kamilakesbi) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:2005]: Actually, `test_flash_attn_2_inference_equivalence` is now computed in [test_modelling_common](https://github.com/huggingface/transformers/blob/8c00b53eb01051e93942e99912455883b7ea47b2/tests/test_modeling_common.py#L3248), so we can remove it here.

**Review by sanchit-gandhi — COMMENTED** [review]
Code changes for Wav2Vec2 look good! A few TODOs before we merge this PR:
- [ ] Update the slow test to confirm the batch inputs give correctness with an attention mask
- [ ] Propagate the changes made to Wav2Vec2 to other models in the library: models like HuBERT get a non-negligible amount of usage (a few hundred thousand downloads per month). It would be good to add support for SDPA and FA2 for all the Wav2Vec2-derived models as well in this PR, such that they get this new feature and code for these models are sync'd properly with the `Copied from` mechanism. No need to add any additional slow tests - if we confirm Wav2Vec2 is correct, and assume that the other models copy from Wav2Vec2, then we can be pretty confident these models are correct as well

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (kamilakesbi) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:1129]: We already check that we get correctness when using the attention mask with FA2 in [test_flash_attn_2_inference_equivalence](https://github.com/huggingface/transformers/blob/8c00b53eb01051e93942e99912455883b7ea47b2/tests/test_modeling_common.py#L3248) with this line:  
https://github.com/huggingface/transformers/blob/8c00b53eb01051e93942e99912455883b7ea47b2/tests/test_modeling_common.py#L3331

I've run this test both with and without updating the hidden-states and they pass, but they don't pass without the line: `attention_mask = attention_mask if (attention_mask is not None and 0 in attention_mask) else None` used in when  `self._use_flash_attention_2` is True. 

I agree that it seems more logical to mask the hidden states as well to copy the behaviour of the other branch. I'll add that in my next commit.

File: src/transformers/models/wav2vec2/modeling_wav2vec2.py
Comment (kamilakesbi) [reply src/transformers/models/wav2vec2/modeling_wav2vec2.py:2110]: It wasn't the case here... before adding this line I got `RuntimeError: mat1 and mat2 must have the same dtype, but got Half and BFloat16`  when running [test_eager_matches_sdpa_inference](https://github.com/huggingface/transformers/blob/8c00b53eb01051e93942e99912455883b7ea47b2/tests/test_modeling_common.py#L3517C9-L3517C42) from `test_modeling_common`.

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (kamilakesbi) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:2007]: Are you sure ? if I remove `torch_dtype=torch.bfloat16` the test fails with the following error: ` RuntimeError: Input type (CUDABFloat16Type) and weight type (torch.cuda.FloatTensor) should be the same`. For now I would still specify it as a parameter.

Comment (kamilakesbi) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:2007]: Are you sure ? when I remove `torch_dtype=torch.bfloat16` I get the following error: `RuntimeError: Input type (CUDABFloat16Type) and weight type (torch.cuda.FloatTensor) should be the same`. For now I'll still specify torch_dtype as a parameter.

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (sanchit-gandhi) [inline tests/models/wav2vec2/test_modeling_wav2vec2.py:2041]: Let's remember to pass the attention mask for batched inputs:

```suggestion
        inputs = processor(input_speech, return_tensors="pt", padding=True, return_attention_mask=True)
        inputs = inputs.to(torch_device)

        with torch.no_grad():
            logits = model_fa(inputs.input_values.to(torch.bfloat16), attention_mask=inputs.attention_mask).logits
```

**Review by sanchit-gandhi — APPROVED** [review]
Looks great! Could you just double check the result of the slow tests with FA2 matches the results without when we pass the attention mask?

Otherwise this PR is ready for core-maintainer review 👍

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (kamilakesbi) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:2041]: I just did. The test passes with the added attention mask ;)

**Comment by sanchit-gandhi** [comment]
Hey @amyeroberts! Would appreciate a final review here when you get the chance - should be a pretty fast PR to review since we leverage lots of `# Copied from` logic! Thanks 🤗

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! Overall looks very clean 

Three main comments: 
* All the models should have FA2 and SDPA tests added to make sure the values are similar to their eager equivalents
* All the models should have FA2 and SDPA info added to the model pages, including an expected expeced performance graph e.g. [like here ](https://huggingface.co/docs/transformers/v4.39.3/en/model_doc/gpt_neox#using-flash-attention-2)
* Some comments in-line about managing the copied from comments

File: src/transformers/models/data2vec/modeling_data2vec_audio.py
Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:501]: Note that the attention module in Wav2Vec2 is one-to-one the same as BART (self-attn and cross-attn), but inherently different from LLaMA (self-attn only). Therefore, we copy the main attention class from BART, and only override the FA2 forward method from LLaMA. This is consistent with how we implement FA2 in Whisper:
https://github.com/huggingface/transformers/blob/60dea593edd0b94ee15dc3917900b26e3acfbbee/src/transformers/models/whisper/modeling_whisper.py#L398-L399

I would be in favour of maintaining consistency with both Whisper, and the non-FA2 attention class, where we copy from BART and only override the specific FA2 methods that come from LLaMA.

**Comment by sanchit-gandhi** [comment]
That's a great point regarding the model docs - would you like to run a quick benchmark for Wav2Vec2 and HuBERT @kamilakesbi and subsequently update the respective model docs?

<details>

<summary> You can use the following code snippet as a starting point </summary>
```python
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor
from datasets import load_dataset
from evaluate import load
import torch
from tqdm import tqdm

# define our torch configuration
device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "distil-whisper/distil-large-v3"

# load the model + processor
model =  AutoModelForSpeechSeq2Seq.from_pretrained(model_id, torch_dtype=torch_dtype, use_safetensors=True, low_cpu_mem_usage=True)
model = model.to(device)
processor = AutoProcessor.from_pretrained(model_id)

# load the dataset with streaming mode
dataset = load_dataset("librispeech_asr", "clean", split="validation")

# define the evaluation metric
wer_metric = load("wer")

def inference(batch):
    # 1. Pre-process the audio data to log-mel spectrogram inputs
    audio = [sample["array"] for sample in batch["audio"]]
    input_features = processor(audio, sampling_rate=batch["audio"][0]["sampling_rate"], return_tensors="pt").input_features
    input_features = input_features.to(device, dtype=torch_dtype)
    
    # 2. Auto-regressively generate the predicted token ids
    pred_ids = model.generate(input_features, max_new_tokens=128)
    
    # 3. Decode the token ids to the final transcription
    batch["transcription"] = processor.batch_decode(pred_ids, skip_special_tokens=True)
    batch["reference"] = batch["text"]
    return batch

# batch size 16 inference
dataset = dataset.map(function=inference, batched=True, batch_size=16)

# normalize predictions and references
all_transcriptions = [processor.normalize(transcription) for transcription in dataset["transcription"]]
all_references = [processor.normalize(reference) for reference in dataset["reference"]]

# compute the WER metric
wer = 100 * wer_metric.compute(predictions=all_transcriptions, references=all_references)
print(wer)
```

</details>

File: src/transformers/models/data2vec/modeling_data2vec_audio.py
Comment (amyeroberts) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:501]: OK - let's keep things consistent!

File: tests/models/wav2vec2/test_modeling_wav2vec2.py
Comment (amyeroberts) [reply tests/models/wav2vec2/test_modeling_wav2vec2.py:1]: OK - if the models aren't used much then let's leave it!

**Comment by kamilakesbi** [comment]
I've run a quick benchmark on both [Wav2Vec2](https://huggingface.co/datasets/kamilakesbi/transformers_image_doc/resolve/main/data/Wav2Vec2_speedup.png) and [Hubert](https://cdn-lfs-us-1.huggingface.co/repos/69/17/691746035d62c6bde74a208d3008beb03c0389cb739ba4d63de499dd5f5fcf5a/1b0112bb18bc8ac94db2f937a518e22bd89554b6fafdaa157223b45a0267c883?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27Hubert_speedup.png%3B+filename%3D%22Hubert_speedup.png%22%3B&response-content-type=image%2Fpng&Expires=1714062523&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxNDA2MjUyM319LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy11cy0xLmh1Z2dpbmdmYWNlLmNvL3JlcG9zLzY5LzE3LzY5MTc0NjAzNWQ2MmM2YmRlNzRhMjA4ZDMwMDhiZWIwM2MwMzg5Y2I3MzliYTRkNjNkZTQ5OWRkNWY1ZmNmNWEvMWIwMTEyYmIxOGJjOGFjOTRkYjJmOTM3YTUxOGUyMmJkODk1NTRiNmZhZmRhYTE1NzIyM2I0NWEwMjY3Yzg4Mz9yZXNwb25zZS1jb250ZW50LWRpc3Bvc2l0aW9uPSomcmVzcG9uc2UtY29udGVudC10eXBlPSoifV19&Signature=KTFcjDntYyZRTD0IoZFWGumqE2BCErcb5-69r6%7EgH-87V7TRedao4z-x%7E2KZKh5J5wPHKoO2hke1PzIqdIqtMgcZccqMpeLGdZGkUP9ryAwMMM-Q%7EX003sFzC79P2wE25TW27I4O1Q5YRmVG3n%7E5Dv4QTGTm307o6GigM03QeyHEEOEKyVAjx3TlpPd9Hotj6qySVd2Lqn5PU82JlNGrUmZaVifz49kcb3aRX8lAQs2Diu2czLB6P3rZZs5JTL88JC3N-aha90%7EsCNQEoB-fuy3AuiN1offF0KFnEEg3QxI5hwyHz0JUmr9Z%7EwMvfsFqdB4%7EEjaYQiu5h48RLXk7kg__&Key-Pair-Id=KCD77M1F0VK2B) and updated the doc. 

I think we can merge this PR @amyeroberts if you validate the plots :)

**Review by amyeroberts — APPROVED** [review]
Looks great! Thanks for running the FA2 and SDPA speedup comparisons 🚀

**Pull request merged**
